### PR TITLE
Add more easy accessors for Exif & TIFF/EP (0.27->master)

### LINKED ISSUE #1054: Problem with Huawei P30 Pro camera data from dng file
Exiv2 version 0.25 doesn't read the correct camera data from the Huawei P30 Pro dng file.  No problem reading data from the jpg file. 

ISO value is correct, but aperture and exposure are set to zero.

Camera data from the Huawei P20 Pro dng file is correct.

The zip file contains a dng file from a Huawei P30 Pro.

Thank you,
Don

[IMG_20191007_114345.zip](https://github.com/Exiv2/exiv2/files/3813349/IMG_20191007_114345.zip)

### LINKED ISSUE #1385: Complete easy access coverage for Exif & TIFF/EP
**Describe the solution you'd like**
There are a few more tags that overlap between Exif and TIFF/EP that are not covered in src/easyaccess.cpp (e.g. `ShutterSpeedValue`, `ApertureValue`, etc.)

It would be great to complete the coverage to avoid users having to manually check both Exif.Photo and Exif.Image IFDs.

This would help improve support of DNGs for example.

### PR DESCRIPTION
**Pull request by kmilos**
Fixes https://github.com/Exiv2/exiv2/issues/1385

Fixes #1054

### CODE REVIEW

**Comment by clanmills** [comment]
Thank You, @kmilos I'll review this later.  At first glance this looks good.  You've added several functions to the API which is OK and as you've included Comments, they will be documented by Doxygen.

I wonder if there's a need for a document to explain the purpose, use and design of the "easyaccess" APIs.  Something for the wiki perhaps?

**Comment by clanmills** [comment]
Most of the discussion about this matter is here:  #1385

**Review by LeoHsiao1 — DISMISSED** [review]
It looks like a small change. What do I need to do? @clanmills

**Comment by clanmills** [comment]
@LeoHsiao1.  The changes in samples/easyaccess-test.cpp are small at present.  I'd like you to port test/iso65k-test.sh to tests/bashtests.

It's highly likely that iso65k-test.sh will need additional code to fully exercise changes in this PR.  So, I'm asking to carry out a two step project:

1) port test/iso65k-test.sh to test/bashtests
2) In a few weeks there may be changes to iso65k-test.sh to be ported. 

I've never used the easyacess API and there's very little explanation in the code.  Now is a good time to explore this API and get it documented in the book and the Wiki.  The exiv2 command-line application option `-ps` reports a summary of the metadata and some is obtained using the easyaccess API.  I suspect that all data in `-ps` could be reported using the easyaccess API with additional changes to easyaccess.  @kmilos has not committed to doing that:  https://github.com/Exiv2/exiv2/issues/1385#issuecomment-721088495

This also has something to do with DNG (Adobe's Digital Negative Specification).  Draft here.  http://www.barrypearson.co.uk/top2009/downloads.htm

I think we've stumbled into something in the code base that's useful and almost undocumented.

Can you port test/iso65k-test.sh to test/bashtests while I investigate `exiv2 -ps` and DNG.

It's very likely that iso65k-test.sh will be extended by my investigation.  Once, I fully understand easyaccess, I have agreed to document it in both the book and on the wiki.

**Comment by LeoHsiao1** [comment]
I've just seen an old love movie. Now I don't see what you mean. @clanmills 
The last modification time of the file `exiv2/test/iso65k-test.sh` was 26 Sep 2019. And it has already been rewritten to `exiv2/tests/bash_tests/testcases.py` last month. So, should I wait until `iso65k-test.sh` is modified?
In addition, should I submit the modifications together to this PR?

**Comment by clanmills** [comment]
You are quite right, @LeoHsiao1  You can go back to your old movies (see photo below)

@kmilos Please update your PR when you are ready.  I'll review/merge it today.  

Very good team work by everybody.  Two days ago, I knew nothing about the easyaccess API.   Now we've documented, extended it, and used it more effectively in src/actions.cpp.

------------------------

tests/bash_tests/testcases.py already has iso65k_test()

```bash
1314 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ grep def tests/bash_tests/testcases.py | grep -i iso
    def iso65k_test(self):
1315 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```

And it's working:

```bash
1317 rmills@rmillsmbp:~/gnu/github/exiv2/kmilos/tests $ python3 runner.py --verbose bash_tests/testcases.py 
addmoddel_test (testcases.TestCases) ... ok
conversions_test (testcases.TestCases) ... ok
crw_test (testcases.TestCases) ... ok
exifdata_test (testcases.TestCases) ... ok
exiv2_test (testcases.TestCases) ... ok
geotag_test (testcases.TestCases) ... ok
icc_test (testcases.TestCases) ... ok
image_test (testcases.TestCases) ... ok
io_test (testcases.TestCases) ... Serving HTTP on 127.0.0.1 port 12760 (http://127.0.0.1:12760/) ...
127.0.0.1 - - [04/Nov/2020 14:35:53] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:53] "HEAD /table.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:53] "GET /table.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:53] "HEAD /table.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:53] "GET /table.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:53] "HEAD /Reagan.tiff HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:53] "GET /Reagan.tiff HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "HEAD /Reagan.tiff HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "GET /Reagan.tiff HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "HEAD /exiv2-bug922a.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "GET /exiv2-bug922a.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "HEAD /exiv2-bug922a.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "GET /exiv2-bug922a.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "HEAD /table.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "GET /table.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "HEAD /table.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "GET /table.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "HEAD /table.jpg HTTP/1.0" 200 -
127.0.0.1 - - [04/Nov/2020 14:35:54] "GET /table.jpg HTTP/1.0" 200 -
ok
iptc_test (testcases.TestCases) ... ok
iso65k_test (testcases.TestCases) ... ok   <----- Good Job
modify_test (testcases.TestCases) ... ok
nls_test (testcases.TestCases) ... ok
path_test (testcases.TestCases) ... ok
png_test (testcases.TestCases) ... ok
preview_test (testcases.TestCases) ... ok
stdin_test (testcases.TestCases) ... ok
stringto_test (testcases.TestCases) ... ok
tiff_test (testcases.TestCases) ... ok
version_test (testcases.TestCases) ... ok
webp_test (testcases.TestCases) ... ok
write2_test (testcases.TestCases) ... ok
write_test (testcases.TestCases) ... ok
xmpparser_test (testcases.TestCases) ... ok

----------------------------------------------------------------------
Ran 24 tests in 21.845s

OK
1318 rmills@rmillsmbp:~/gnu/github/exiv2/kmilos/tests $ 
```

--------------------
On Monday, we completed the installation of a Samsung QLED 65" (2000mm) TV at home.  Fully connected to the Internet, TV, streaming, games, music and more.  Better than the cinema.

img_1919

**Comment by kmilos** [comment]
Nice setup! I'm still daydreaming of a home projector, but need to keep ambient light under control for those...

Anyhow, the PR is now I think ready for final review.

File: src/actions.cpp
Comment (clanmills) [inline src/actions.cpp:338]: This code is tough to read/understand
```cpp
        // From ExposureTime, failing that, try ShutterSpeedValue
        printLabel(_("Exposure time"));
        bool done = 0 != printTag(exifData, Exiv2::exposureTime);   // do we mean done = printTag(....) != 0
        if (!done) {
            done = 0 != printTag(exifData, Exiv2::shutterSpeedValue); // # why are we setting done?
        }
        std::cout << std::endl;
        // Get if from FNumber and, failing that, try ApertureValue
        printLabel(_("Aperture"));
        done = 0 != printTag(exifData, Exiv2::fNumber);
        if (!done) {
            done = 0 != printTag(exifData, Exiv2::apertureValue); # why set done?
        }
        std::cout << std::endl;
```

Can we "hide" the details in another printTag() method which takes two easyaccess function selectors?  Then 14 lines of if statements can be expressed in two lines that match the pattern of code like this:

```cpp
printTag(exifData, Exiv2::exposureTime, Exiv2::shutterSpeedValue, _("Exposure time"));
printTag(exifData, Exiv2::fNumber     , Exiv2::apertureValue    , _("Aperture"));
```

When you add `printTag(exifData, Exiv2::fNumber     , Exiv2::apertureValue    , _("Aperture"));`, can you add spaces to the printTag's above and below, so that the whole thing is instantly readable.  Something like:

```cpp
        printTag(exifData, Exiv2::dateTimeOriginal                                , _("Image timestamp"));
        printTag(exifData, "Exif.Canon.FileNumber"                                , _("File number"));
        printTag(exifData, Exiv2::exposureTime          , Exiv2::shutterSpeedValue, _("Exposure time"));
        printTag(exifData, Exiv2::fNumber               , Exiv2::apertureValue    , _("Aperture"));
        printTag(exifData, Exiv2::exposureBiasValue                               , _("Exposure bias"));
        printTag(exifData, Exiv2::flash                                           , _("Flash"));
```

**Review by clanmills — COMMENTED** [review]
I've made a suggestion to make Print::printSummary() easier to read.

**Comment by clanmills** [comment]
We discussed a home projector.  The big win is that you don't have a big black blob on the wall when it's not in use.

We decided to get the QLED TV because it has a "ambient" mode for showing artwork or photos.  We got the local framing shop to make an oak surround so it looks quite nice when not in use.  TVs are developing as fast as phones and computers.  The TV has wifi, browser, games, streaming, displays photos, connects to Spotify and all manner of stuff.

We decided not to get a FaceBook Portal (Webcam) which could go above the sound bar.

File: src/actions.cpp
Comment (kmilos) [reply src/actions.cpp:338]: Ok, I'll clean this up as soon as I get a chance. Maybe with a slight variation to avoid addition of yet another printTag() overload, so how about this single prototype instead?

``` cpp
int printTag(const Exiv2::ExifData& exifData,
                     EasyAccessFct easyAccessFct,
                     const std::string& label = "",
                     EasyAccessFct easyAccessFctFallback = NULL) const;
```

Then we can have e.g.:

``` cpp
        printTag(exifData, Exiv2::fNumber,            _("Aperture"),     Exiv2::apertureValue);
        printTag(exifData, Exiv2::exposureBiasValue,  _("Exposure bias") );
```

Otherwise I can put the fallback function argument before the label, but then I have to explicitly specify a lot of `NULL`s everywhere else... It's a style/readability preference:

``` cpp
        printTag(exifData, Exiv2::fNumber,           Exiv2::apertureValue, _("Aperture"));
        printTag(exifData, Exiv2::exposureBiasValue, NULL,                 _("Exposure bias"));
```

Flip a coin ;)

**Comment by kmilos** [comment]
> We decided not to get a FaceBook Portal (Webcam) which could go above the sound bar.

Probably wise, given all security and privacy concerns in general, especially if the camera doesn't have a mechanical cover...

Anyhow, let me know about the comment feedback, I'll finally implement whatever you prefer out of the 3 options (your initial, one or my 2 variations).

File: src/actions.cpp
Comment (clanmills) [reply src/actions.cpp:338]: Sure, good suggestions.  I like the optional fourth "fallback selector" parameter.

Now you see the benefit of having Team Exiv2 write access to our repos.  I could updated your PR and ask you to approve it.  Simpler work-flow.

**Comment by clanmills** [comment]
I've commented on your good suggestions.  Thank You for your suggestions.

The Facebook Portal has a "sliding hood" which can cover the lens.  We have several "Google Home" devices.  I'm sure the FBI/KGB are listening and watching us.  If you hear of an agent dying of boredom, she's the person assigned to our surveillance.  I also hope she's deaf and doesn't need to listen to me practicing the euphonium!

MusicRoom

**Review by clanmills — APPROVED** [review]
Beautifully executed.  Excellent work.  Thank You very much.

**Comment by clanmills** [comment]
I'll merge this when the CI completes provided it's happy!

**Comment by kmilos** [comment]
Thanks, pleasure.

When you finally get the time to watch, add [this film](https://www.imdb.com/title/tt0339535) to your list (a precursor to the more acclaimed German "The Lives of Others) ;)

**Comment by clanmills** [comment]
This has passed the CI checks.  Time to merge and close.

I'll look at your film after lunch.

I've written a Wiki Page.  https://github.com/Exiv2/exiv2/wiki/EasyAccess-API  _**Comments Welcome**_.

I don't intend to email Carsen about EasyAccess.  I think we've beaten this to death.  Time to move on!

**Pull request merged**
